### DISCUSSION TITLE
How to register multiple OAuth logins to the same user?

### DISCUSSION THREAD

**Discussion by oOBoomberOo** [jaredhanson/passport#895] [Help] [answered]
I want the user to be able to register using one of the third-party providers and then have them connect other providers to the existing account instead of registering a new one. What would be the unique identity that tells who the user is in this case?

**Comment by saikarthikp9** [comment]
Did you figure this out? I guess if the email IDs match you can assume they are indeed the same user.

**Reply by oOBoomberOo** [reply]
Yes, I finally figured it out. You can set `passReqToCallback` to true in the strategy's option, which will make the first argument of the verifier function be the Request object. Then you can look up `req.user?` to see if the user is currently authenticated through another strategy or not and handle it accordingly.

```js
new TwitterStrategy({
    callbackURL: "...",
    consumerKey: "...",
    consumerSecret: "...",
    passReqToCallback: true
}, (req, accessToken, refreshToken, profile, done) => {
    const hasAlreadyRegistered = hasRegistered(profile.id); // `id` is a unique ID from the platform that the user is authenticating. You can use it to look up if this account has been signed up before.
    if (req.user && !hasAlreadyRegistered) {
        // add this provider to the existing user
    }
    else if (hasAlreadyRegistered) {
        // find the existing user with that `profile.id` and return it
    }
    else {
        // create a new user and then add the provider
    }
})
```

**Comment by oOBoomberOo** [comment, marked as answer]
The `passReqToCallback` option can be used to pass the Request object to the verifier function, and then you can handle the connecting account logic by checking if the `req.user` object exists or not.
